project/src/main/kotlin/org/ionproject/core/accessControl/Cache.kt

```diff
@@ -0,0 +1,106 @@
+package org.ionproject.core.accessControl
+
+import org.ionproject.core.accessControl.pap.entities.PolicyEntity
+import org.ionproject.core.accessControl.pap.entities.TokenEntity
+import org.ionproject.core.accessControl.pap.sql.AuthRepoImpl
+import org.ionproject.core.common.LogMessages
+import org.ionproject.core.common.customExceptions.BadRequestException
+import org.ionproject.core.common.customExceptions.UnauthenticatedUserException
+import org.ionproject.core.common.interceptors.LoggerInterceptor
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+
+/**
+ *  Cache used to avoid having to check database everytime
+ *  a request is made and needs authentication.
+ */
+@Component
+class AccessControlCache(private val pap: AuthRepoImpl, cacheConfig: CaffeineConfiguration) {
+
+    private val tokenCache = cacheConfig.tokenCache
+    private val policiesCache = cacheConfig.policiesCache
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(LoggerInterceptor::class.java)
+    }
+
+    /**
+     * saves the requested token by the hash for future
+     * requests.
+     */
+    fun getToken(tokenHash: String, derived: Boolean) : TokenEntity {
+        return tokenCache.asMap().compute(tokenHash) { _, oldValue ->
+            {
+                if (oldValue == null) {
+                    logger.info("No token cache hit, reading token from database...")
+
+                    val token: TokenEntity = readTokenDb(tokenHash, derived)
+                    token
+                } else {
+                    logger.info("Token cache hit!")
+                    oldValue
+                }
+            }()
+        }
```

Comment:
fyi,
This forces the entry to be locked during a compute in order to be atomic, which degrades the read throughput as it becomes sequential for a given key. It is still reasonably fast (40-60M ops/s). Generally prefer using a `computeIfAbsent` call, either via `cache.get` or `asMap`, which will be a lock-free read on a cache hit. You might prefer enabling and expose stats rather than logging.

Reply:
I see, thanks for the information, I'll update the code to use Cache#get(key, mappingFunction).